#!/bin/sh
set -eu

if ! command -V patch >/dev/null 2>&1; then
  extra_root="$(readlink -f "$(dirname -- "$0")/..")"
  "${extra_root}/bin/pkg" add patch
fi

generator="$(find /usr/lib/python*.*/site-packages/batocera_launch_dolphin -name controllers.py)"
patch -b -z .backup-cemuhook "${generator}" << 'EOF'
--- "${generator}"
+++ controllers.py
@@ -65,6 +65,14 @@
         else:
             generate_controller_config_realwiimotes('WiimoteNew.ini', 'Wiimote')
             generate_controller_config_gamecube(emulator, players_controllers, {}, rom)
+        if 'cemuhook_server_port' in emulator.config:
+            config_file = DOLPHIN_CONFIG / "DSUClient.ini"
+            with config_file.open('w', encoding='utf_8_sig') as f:
+                cemuhook_server_ip = emulator.config.get('cemuhook_server_ip')
+                cemuhook_server_port = emulator.config.get('cemuhook_server_port')
+                f.write(f'[Server]\n')
+                f.write(f'Enabled = True\n')
+                f.write(f'Entries = batocera-extra:{cemuhook_server_ip}:{cemuhook_server_port};\n')
     elif emulator.system == 'gamecube':
         used_wheels: DeviceInfoMapping = {}
         if (
@@ -731,6 +739,20 @@
             f.write(f'IMUAccelerometer/Backward = `{dev}:Accel Z+`\n')
             f.write(f'IMUAccelerometer/Up = `{dev}:Accel Y-`\n')
             f.write(f'IMUAccelerometer/Down = `{dev}:Accel Y+`\n')
+        # Cemuhook motion control
+        if 'cemuhook_server_port' in emulator.config:
+            f.write(f'IMUGyroscope/Pitch Up = `DSUClient/0/batocera-extra:Gyro Pitch Up`\n')
+            f.write(f'IMUGyroscope/Pitch Down = `DSUClient/0/batocera-extra:Gyro Pitch Down`\n')
+            f.write(f'IMUGyroscope/Roll Left = `DSUClient/0/batocera-extra:Gyro Roll Left`\n')
+            f.write(f'IMUGyroscope/Roll Right = `DSUClient/0/batocera-extra:Gyro Roll Right`\n')
+            f.write(f'IMUGyroscope/Yaw Left = `DSUClient/0/batocera-extra:Gyro Yaw Left`\n')
+            f.write(f'IMUGyroscope/Yaw Right = `DSUClient/0/batocera-extra:Gyro Yaw Right`\n')
+            f.write(f'IMUAccelerometer/Left = `DSUClient/0/batocera-extra:Accel Left`\n')
+            f.write(f'IMUAccelerometer/Right = `DSUClient/0/batocera-extra:Accel Right`\n')
+            f.write(f'IMUAccelerometer/Forward = `DSUClient/0/batocera-extra:Accel Forward`\n')
+            f.write(f'IMUAccelerometer/Backward = `DSUClient/0/batocera-extra:Accel Backward`\n')
+            f.write(f'IMUAccelerometer/Up = `DSUClient/0/batocera-extra:Accel Up`\n')
+            f.write(f'IMUAccelerometer/Down = `DSUClient/0/batocera-extra:Accel Down`\n')
         # Mouse to emulate Wiimote
         if emulator.config.get_bool('mouseir'):
             f.write('IR/Up = `Cursor Y-`\n')
EOF

echo >&2 "Dolphin generator was patched for cemuhook."
